components/raftstore/src/store/entry_storage.rs

```diff
@@ -980,8 +1073,128 @@ impl<ER: RaftEngine> EntryStorage<ER> {
+    }
+
+    /// Warm up entry cache if the result is valid.
+    ///
+    /// Return true when the warmup operation succeed within the timeout.
+    pub fn maybe_warm_up_entry_cache(&mut self, res: RaftlogFetchResult) -> bool {
+        let low = res.low;
+        // Warm up the entry cache if the low and high index are
+        // exactly the same as the warmup range.
+        let state = self.entry_cache_warmup_state().as_ref().unwrap();
+        let range = state.range();
+        let is_task_timeout = state.is_task_timeout();
+
+        if range.0 != low {
+            return false;
+        }
+
+        match res.ents {
+            Ok(mut entries) => {
+                let last_entry_index = entries.last().map(|e| e.index);
+                if let Some(index) = last_entry_index {
+                    // Generally speaking, when the res.low is the same as the warmup
+                    // range start, the fetch result is exactly used for warmup.
+                    // As the low index of each async_fetch task is different.
+                    // There should exist only one exception. A async fetch task
+                    // with same low index is triggered before the warmup task.
+                    if index + 1 >= range.1 {
+                        let is_valid = if let Some(first_index) = self.entry_cache_first_index() {
+                            range.1 == first_index
+                        } else {
+                            range.1 == self.last_index() + 1
+                        };
+                        assert!(is_valid, "the warmup range should still be valid");
+                        entries.truncate((range.1 - range.0) as usize);
+                        self.cache.prepend(entries);
+                        WARM_UP_ENTRY_CACHE_COUNTER.finished.inc();
+                        fail_point!("on_entry_cache_warmed_up");
+                        return !is_task_timeout;
+                    }
+                }
+                warn!(
+                    "warm up the entry cache failed";
+                    "region_id" => self.region_id,
+                    "peer_id" => self.peer_id,
+                    "last_entry_index" => last_entry_index.unwrap_or(0),
+                    "expected_high" => range.1,
+                );
+            }
+            Err(e) => {
+                warn!(
+                    "warm up the entry cache failed";
+                    "region_id" => self.region_id,
+                    "peer_id" => self.peer_id,
+                    "err" => ?e,
+                );
+            }
+        }
+        false
+    }
+
     pub fn compact_entry_cache(&mut self, idx: u64) {
-        self.cache.compact_to(idx);
+        let mut can_compact = true;
+        if let Some(state) = self.entry_cache_warmup_state_mut() {
+            if state.check_stale(MAX_WARMED_UP_CACHE_KEEP_TIME) {
+                self.clear_entry_cache_warmup_state();
+            } else {
+                can_compact = false;
```

Comment:
This looks suspicious (in triggering OOM). Is it really needed?

Reply:
Before, a peer loads entries into cache after it becomes leader. Now, entires are pre-fetched for a peer which is about to become leader. So in theory, the memory consumption should be similar.

If a peer has entry_cache_warmup_state, it means this peer is ready to become leader and it has a great possibility to become leader.

Reply:
> loads entries into cache after it becomes leader

Could you point me to the code? I presume it's async? i.e. can it still become leader if entries are not all in cache.

Reply:
> > loads entries into cache after it becomes leader
> 
> Could you point me to the code? I presume it's async? i.e. can it still become leader if entries are not all in cache.

When a leader replicates it's entries to other followers which has lagging logs, it fetch entries from raftdb (by `async_fetch`) and load them into entry_cache. 
https://github.com/tikv/tikv/blob/7fbbcdcee6f7381a9c048030c8f95fac055c42bb/components/raftstore/src/store/entry_storage.rs#L843-L844

Reply:
Ignore me, as long as `evict_entry_cache` is not changed, the OOM situation should be fine.